---
title: "GTExExploration"
author: "Nick Burns"
date: "15 April 2016"
output: html_document
---

# Analysis of GTEx data  

This project began as a simple way to tinker with R's data.table package using gene expression data from GTEx. However, along the way we found some really interesting challenges with the data. This notebook will extend on the initial exploration and begin to tackle some of the interesting problems.

## Read in the data  

NOTE: this is the reduced 40-gene set

```{r}
setwd("C:\\Users\\NickBurns\\gitRepositories\\myGits\\GTExEDA")
#setwd("~/Documents/GitHub/GTExEDA")
#dataDir <- "/mnt/DataDrive/GTEXData/%s"
dataDir <- "C:/Users/NickBurns/Downloads/%s"

library(data.table)

load("targetGenes.Rdata")
load("GTEx_40Genes.Rdata")

dictFile <- sprintf(dataDir, "GTEx_Data_V6_Annotations_SampleAttributesDS.txt")
dict <- fread(dictFile)
dict <- dict[, .(SAMPID, SMTS, SMTSD)]
```

## Tissue imbalance  

Something that immediately jumped out at me, was the huge imbalance in the number of samples per tissue (see plot below)

```{r}
library(ggplot2)

freqTissues <- table(dict[, SMTS])
tissueOrder <- names(freqTissues[order(freqTissues, decreasing = FALSE)])
dict[, SMTS := factor(.SD[,SMTS], levels = tissueOrder, ordered = TRUE)]

ggplot(dict, aes(x = SMTS)) + 
    geom_bar(fill = "steelblue") + 
    coord_flip() +
    theme_bw()

```

Clearly, tissues such as blood, brain and skin are represented significantly more than Bone marrow, prostate or spleen. Of course we know that most statistical optimisation methods rely on the classes being balanced. So I suspect that this imbalance will be affecting the quality of results. 

So my first challenge will be to cluster the genes (will simply use heirarchical clustering) and then perform feature selection on the cluster IDS. The goal is to understand which tissues are driving the cluster membership and perhaps use this as a form of dimensionality reduction.  

### Cluster the genes  

```{r}
idxExpressionCols <- which(colnames(myGTEx) %in% colnames(myGTEx)[-c(1:2)])
model <- hclust(dist(myGTEx[, idxExpressionCols, with = FALSE][, log2(.SD + 1)]))
plot(model, labels = myGTEx[, Description])

myGTEx[, ClusterIDs := factor(cutree(model, k = 8))]
head(myGTEx[, c(1:5, ncol(myGTEx)), with = FALSE])
```

### Use a random forest to predict ClusterID  

```{r}
library(randomForest)
forest <- randomForest(myGTEx[, idxExpressionCols, with = FALSE],
                       y = myGTEx[, ClusterIDs],
                       importance = TRUE)
```

And a quick look at the training accuracy:  

```{r}
with(forest, table(predicted, myGTEx[, ClusterIDs]))
```

Now, let's identify the top 50 tissue samples:

```{r}
features <- forest$importance[order(forest$importance[, 2],
                                    decreasing = TRUE), 2]
head(features, 50)
```

Nice, it has actually done quite a good job. What now if we build the model with only the top 10 tissues?

```{r}
f2 <- randomForest(myGTEx[, which(colnames(myGTEx) %in% names(features)[1:10]), with = F],
                   y = myGTEx[, ClusterIDs])
table(f2$predicted, myGTEx[, ClusterIDs])
```

This is really interesting - the random forest is able to do very well with only the top 10 features. I know that the training accuracy is a very poor indicator however. So I need to:  
  1. Use k-fold cross validation to confirm the accuracy  
  2. Visualise somehow the way that these tissues are contributing to the prediction of gene clusters (I suspect that only 1 or 2 tissues are dominating a cluster, and this might be a tentative guide to differential expression (at least in a predictve sense)).  
  3. PLAY!
  
  
Let's see if the top 100 features can yield similar gene loadings via factor analysis:

```{r}
getFeatures <- function (K) {
    which(colnames(myGTEx) %in% names(features)[1:K])
}
fData <- myGTEx[, getFeatures(4000), with = FALSE][, t(.SD)]
colnames(fData) <- myGTEx[, Description]
head(fData[, 1:10])

fit <- factanal(fData, factors = 7)


NGenes <- 40
plotGeneLoadings <- data.frame(Z = rep(1:7, each = NGenes), 
                               loading = fit$loadings[1:(7 * NGenes)], 
                               gene = rep(myGTEx[, Description], 7))

ggplot(plotGeneLoadings, aes(x = Z, y = loading)) +
    geom_text(aes(label = gene, size = loading**2)) +
    theme_bw() +
    ggtitle("Strength of gene loading by factor") +
    theme(legend.position = "none")
```

2600 (K = 2600) was the fewest features that would result in convergence. At this level, the clusters were definitely messy. At K = 4000, the clusters were sensible again. This is interesting, it indicates that we can throw out approx. half the tissue samples (not sure which ones...), but that there is a fair level of complexity to the correlation in gene expressions necessary to distinguish the clusters. 